src/server.rs

```diff
@@ -127,29 +131,25 @@ macro_rules! messages {
                 });
             }
 
-            let id: Option<usize> = match ls_command.get("id") {
-                Some(v) => {
-                    if v.is_u64() {
-                        Some(v.as_u64().unwrap() as usize)
-                    } else if v.is_string() {
-                        match usize::from_str_radix(v.as_str().unwrap(), 10) {
-                            Ok(v) => Some(v),
-                            Err(_) => None
-                        }
-                    } else {
-                        None
-                    }
-                },
-                None => None
+            let id = ls_command.get("id").expect("no id").to_owned();
+            let id: jsonrpc::Id = serde_json::from_value(id).expect("Can't extract id");
```

Comment:
Good catch! Missed that. Added handling missing id, but I now realized that according to JSON-RPC/LSP failure responses are only sent as a response to requests (and not notifications).

I'd say we should eventually change return value of `parse_message` to `Result<ServerMessage, Option<jsonrpc::Failure>>` and handle/output failures only for received invalid Requests, but with this we're not making the flow any worse wrt that and we can improve it the next iteration. Added few reminding comments. @nrc what do you think?